### PR TITLE
Add country switch component for WC Payments NOX

### LINKED ISSUE #52665: [WC Payments NOX] Add country switch component
### Description

For the business location control, we don't use a simple dropdown; instead, we use a dedicated React component with search functionality and a confirmation modal.

Dependency for #51454 

### Acceptance criteria

* The business location control uses a React component following the designs (with the optional Apply button).

### Designs
Design post: p6riRB-bfm-p2
Figma for full solution: xO8YdpKfaVoP2d8NvweZ5I-fi-3297_85670
Figma for confirmation modal: xO8YdpKfaVoP2d8NvweZ5I-fi-3297_86200
Figma for alternative solution without Apply button: xO8YdpKfaVoP2d8NvweZ5I-fi-3297_85935

### Dev notes
There is [this WooPayments PR](https://github.com/Automattic/woocommerce-payments/pull/8922) with a similar component.

Reactifying the Payments settings page: pe4Hk0-1ly-p2
Populate react-powered main payments settings screen PR: https://github.com/woocommerce/woocommerce/pull/50825

### Additional context
PRD: 1wMfX8CAjDnOp24tcU-PADB0pOt3cz7mGdyZYeRBxx2Y-gdoc

cc @timmy5685 @mordeth

### PR DESCRIPTION
**Pull request by mordeth**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #52665.

This PR introduces a React component for the business location control, featuring a search function and an “Apply” button to confirm the selected country.

Screenshot 2024-11-22 at 10 06 48

Demo: https://www.loom.com/share/1e18ec3508f041de8f02dac80ff70c08?sid=e85365fd-3704-4573-9cb8-3d101bd91330

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up a WooCommerce test site on JN using this branch and install the WC Beta Tester plugin.
2. In WC Beta Tester, go to the Features tab and enable the “reactify-classic-payments-settings” feature flag.
3. Complete the WooCommerce setup using the Core Profiler. You can skip or complete the full setup—just note the country selected, as it will be important later.
4. Navigate to ./wp-admin/admin.php?page=wc-settings&tab=checkout.
5. Verify that the dropdown displays the correct business location selected during setup.
6. Test the dropdown functionality to ensure it works as expected.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/admin/package.json [hunk 1]
```diff
@@ -81,9 +81,11 @@
 		"@wordpress/warning": "wp-6.0",
 		"@xstate/react": "3.2.1",
 		"@xstate5/react": "npm:@xstate/react@4",
+		"classnames": "^2.3.2",
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/package.json]: clsx provides similar functionality to classnames, and we've got a direction to migrate to clsx: pdnLyh-6ic-p2

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx [hunk 2]
```diff
@@ -31,6 +32,10 @@ export const PaymentGateways = ( {
 	setupPlugin,
 }: PaymentGatewaysProps ) => {
 	const setupLivePayments = () => {};
+	const [ storeCountry, setStoreCountry ] = useState( 'US' );
+	const countryOptions = Object.entries( window.wcSettings.countries || [] )
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx]: What do you think about memoizing this since it's a pretty long list that doesn't change?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+			// translators: %s: The selected option.
+			__( 'Currently selected: %s', 'woocommerce' ),
+			itemString
+		);
+	}
+
+	const getSearchSuffix = ( focused: boolean ) => {
+		if ( focused ) {
+			return (
+				<svg
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Would you consider moving the svg out either into a .svg asset or into another named react component if that's not a good option? Having this svg inline makes it hard to read and also doesn't give an idea of what image it represents

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	const onApplyHandler = useCallback(
+		( e: React.MouseEvent< HTMLButtonElement > ) => {
+			e.stopPropagation();
+			onChange( selectedValue );
+			closeMenu();
+		},
+		[ onChange, selectedValue, closeMenu ]
+	);
+
+	const handleSearch = ( {
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: I'd debounce this for the fast typists 😄

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+		target,
+	}: React.ChangeEvent< HTMLInputElement > ) => {
+		if ( ! previousStateRef.current ) {
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Since there's already a pattern of using useReducer here, I think it'd make things clearer and easier for future changes if we also use a reducer for visible items. See https://gist.github.com/rjchow/52497fcb5bfbd43bc8e64f935600b05a for a rough idea of what I have in mind

**Review by rjchow — COMMENTED** [review]
Hi! Thanks for looping us in on the review - I've left some comments.

### Re: PR Testing instructions
Usually these instructions are written not just for devs reviewing but also for globalstep testers, and they're not expected to have a dev environment set up.

So I'd write something like...
1. Set up a WooCommerce test site on JN with this branch, and with WC Beta Tester installed
2. Using WC Beta tester, under the Features tab, toggle on the "reactify-classic-payments-settings" feature flag
3. Set up WC using the Core Profiler, it's ok to skip or go through the full thing. Remember which country you select as it will be relevant later
4. Go to "./wp-admin/admin.php?page=wc-settings&tab=checkout" 
5. Observe that there is a dropdown with the correct business location that was previously selected
6. Test around the dropdown and assert that it works as expected

### Re: the component

- The focus highlight looks a bit disjoint:
image

- It's not possible to tab through the item list, which is not good for accessibility
- Searching doesn't fuzzily match when there are accents, see [this](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/tree-select-control/index.js#L331) for an example

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Internal dependencies
+ */
+import { Item } from './types';
+
+// Truncate the option label if it exceeds the max length.
+export const truncateOptionlabel = ( label: string, maxLength: number ) => {
```
Comment (rjchow) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts]: I think it might be cleaner to use css rule `text-overflow: ellipsis` for this

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx (hunk 2 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx]: Great suggestion! I’ve updated the code to include memoization for the list.

File: plugins/woocommerce/client/admin/package.json (hunk 1 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/package.json]: Thank you for pointing that out! I’ve replaced `classnames` with `clsx`.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 3 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Exporting the SVGs as external files is a great idea. Done!

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 4 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: I initially tried to debounce it, but using a ref prop with debounce doesn’t work because the value remains empty due to the delayed state update. Do you have any suggestions to address this?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts (hunk 6 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts]: Good suggestion! I’ve replaced the JS method with `text-overflow: ellipsis`.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 5 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: That’s a very good suggestion! However, I feel it’s unlikely that the component will grow further, and replacing the current state logic with useReducer might introduce unnecessary complexity. WDYT?

**Review by psealock — COMMENTED** [review]
In testing I went directly to Settings > Payments without kicking off the Profiler. The business location was set to USA. I then completed the profiler setting location to New Zealand and returned to the Payments settings. The location was still set to USA. Without looking at the code, maybe a hook on option update could handle this edge case?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 4 above)
Comment (rjchow) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: I've pushed some changes in https://github.com/woocommerce/woocommerce/pull/53075/commits/2fe937d01a811d30687f0b95fd7463b1e8ad862d that addresses the debouncing and the i18n accent handling - feel free to adjust or amend

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 5 above)
Comment (rjchow) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: After looking closer it's not necessary 😅

**Comment by rjchow** [comment]
> In testing I went directly to Settings > Payments without kicking off the Profiler. The business location was set to USA. I then completed the profiler setting location to New Zealand and returned to the Payments settings. The location was still set to USA. Without looking at the code, maybe a hook on option update could handle this edge case?

@psealock I don't think it's hooked up to change the options yet, I thiiiiiink the scope of the issue is just working on the selector component - @mordeth can confirm if I'm mistaken

**Comment by mordeth** [comment]
> @psealock I don't think it's hooked up to change the options yet, I thiiiiiink the scope of the issue is just working on the selector component - @mordeth can confirm if I'm mistaken.

Correct! The scope of this PR is solely the implementation of the Country Selector component. The list was added only to demonstrate the component.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 4 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Thank you for addressing the debouncing and accents, much appreciated!

**Comment by mordeth** [comment]
@rjchow I’ve implemented arrow key navigation and enter key confirmation to enhance accessibility. Do you think this is a good solution?

**Comment by rjchow** [comment]
Hey @mordeth Im AFK these two days but that sounds good!

**Comment by anu-rock** [comment]
@mordeth I tested the country selector on a fresh JN site. Findings:

The country selector component is working as expected on JN. But the styling is different from your PR description screenshot. The Apply button is not dark blue.

image

##### Keyboard navigation testing

* Can I tab my way to the country selector? :white_check_mark:
  * But the focus is not highlighted, a screen reader might catch it but there is no visual cue (eg. a blue/black outline).
* Does pressing space/return on the selector expands the dropdown? :white_check_mark:
* Can I select another option in the dropdown? :x:
  * I can move up and down in the dropdown using arrow keys but it only works when it was expanded via the space/return key. I cannot do it while tabbing through the inner elements like search input and search icon.
  * Pressing space/return on the highlight option simply closes the dropdown.
* Can I tab over to the inner elements (search, options, apply) when the selector is expanded? :x:
* Can I tab over to the options from search? :x:
  * It goes directly to Apply.
* When the search icon has the focus, can I go back to the search input? :x:
* When I have filtered the options by a search term, the search icon changes to X. Can I press space/return on X to clear the filter? :white_check_mark:

**Comment by vladolaru** [comment]
@anu-rock "But the styling is different from your PR description screenshot. The Apply button is not dark blue."

The color should be picked up from the current user color palette (you can change it from Users > Profile). We don't enforce a certain color. That shade of blue is the current default one in WordPress.

**Comment by mordeth** [comment]
>But the styling is different from your PR description screenshot. The Apply button is not dark blue.

Exactly as Vlad mentioned, I’ve adjusted my color palette to align with the designs.

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx [hunk 7]
```diff
@@ -1,10 +1,10 @@
 /**
  * External dependencies
  */
+import { useState } from 'react';
```
Comment (oaratovskyi) [inline plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx]: shouldn't we be using wordpress/element library for that?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+				/>
+			);
+		}
+
+		return (
+			<img
+				src={
+					WC_ASSET_URL + 'images/settings-payments/search-icon.svg'
+				}
+				alt={ __( 'Clear search', 'woocommerce' ) }
```
Comment (oaratovskyi) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: probably should be `Search`

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+								ref={ searchRef }
+								type="text"
+								value={ searchText }
+								onChange={ ( { target } ) =>
+									setSearchText( target.value )
+								}
+								onFocus={ () => setSearchFocused( true ) }
+								onBlur={ () => setSearchFocused( false ) }
+								tabIndex={ -1 }
+								placeholder={ __( 'Search…', 'woocommerce' ) }
```
Comment (oaratovskyi) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: it's just `Search` in design

**Review by oaratovskyi — COMMENTED** [review]
The design is different when the switcher is open - the country (United States) should show only if it's collapsed
image

it looks not good on smaller screen (I mean size of the Business location... text)
image

also, I believe the focus to the search field should be automatic when the user opens the  country switcher and the search icon should change to clear only if something is typed, not just when it's focused without any letters typed

**Comment by mordeth** [comment]
We decided to address the accessibility improvements in a follow-up PR since the country switcher is a blocker for other issues. cc @anu-rock 

> also, I believe the focus to the search field should be automatic when the user opens the country switcher

@oaratovskyi I don’t believe that’s a good idea, as it could compromise accessibility by changing the focus.

> The design is different when the switcher is open - the country (United States) should show only if it's collapsed.

Honestly, clearing the selected option when the dropdown opens feels a bit odd. @orcunomattic, what’s your take on this?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 8 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Are you referring to the icon or the alt text?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 9 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: Good catch! Updated.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 8 above)
Comment (oaratovskyi) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: alt text!

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx (hunk 7 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx]: We use both in the codebase, but it’s better to keep it consistent. I’ve updated the import to use `@wordpress/element`.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx (hunk 8 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx]: I see what you mean, the alt text for both the search and clear icons was the same. It’s now fixed.

**Review by oaratovskyi — CHANGES_REQUESTED** [review]
There's a bug when you pick the country and click Apply - it's not visible if you open the switcher again
image
but it becomes visible after few seconds, maybe something with state update

File: plugins/woocommerce/changelog/add-confirm-select-control [hunk 10]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Add country switch component for WC Payments NOX
```
Comment (anu-rock) [inline plugins/woocommerce/changelog/add-confirm-select-control:4]:
```suggestion
Add country selector component for the new Payment settings page.
```

Comment (anu-rock) [inline plugins/woocommerce/changelog/add-confirm-select-control]:
```suggestion
Add country selector component for the new Payment settings page.
```

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	} );
+
+	it( 'opens the dropdown menu when the button is clicked', () => {
+		render( <CountrySelector { ...defaultProps } /> );
+		const toggleButton = screen.getByRole( 'combobox' );
+
+		fireEvent.click( toggleButton );
+		expect( screen.getByRole( 'listbox' ) ).toBeInTheDocument();
+		expect( screen.getByText( 'Canada' ) ).toBeInTheDocument();
+		expect( screen.getByText( 'France' ) ).toBeInTheDocument();
```
Comment (anu-rock) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx:55]: Is US not expected to be on the list?

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import '@testing-library/jest-dom/extend-expect';
+import { render, screen, fireEvent } from '@testing-library/react';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { CountrySelector } from '../country-selector';
+
+describe( 'CountrySelector', () => {
```
Comment (anu-rock) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx:13]: Nice suite of tests 👌 Later on we could also add accessibility-related tests.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Internal dependencies
+ */
+import { Item } from './types';
+
+// Get the label for the selected option.
+export const getOptionLabel = ( value: string, options: Item[] ) => {
```
Comment (anu-rock) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts]: Do we really need this very simple utility function? It seems like we can directly use its one-liner statement in the calling code.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx
Comment (anu-rock) [inline plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/country-selector.tsx:1]: I won't pretend to have reviewed this file in depth, but nothing obvious stood out to me.

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx [hunk 14]
```diff
@@ -33,6 +33,18 @@ export const PaymentGateways = ( {
 	setupPlugin,
 	isFetching,
 }: PaymentGatewaysProps ) => {
+	const [ storeCountry, setStoreCountry ] = useState( 'US' );
```
Comment (anu-rock) [inline plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx:36]: Nitpick: Perhaps we have a constant for this somewhere? Ignore if not.

File: plugins/woocommerce/package.json [hunk 15]
```diff
@@ -798,10 +798,8 @@
 				"node_modules/@woocommerce/e2e-core-tests/CHANGELOG.md",
 				"node_modules/@woocommerce/api/dist/",
 				"node_modules/@woocommerce/admin-e2e-tests/build",
-				"node_modules/@woocommerce/classic-assets/build",
```
Comment (anu-rock) [inline plugins/woocommerce/package.json]: These removals seem to be unrelated to the code. Ignore if intentional.

**Review by anu-rock — APPROVED** [review]
Code-only review. Pre-approving with minor comments.

___

#### Review checklist

| Parameter  | Result |
| ------------- | ------------- |
| Approach  | ✅ All good |
| Implementation  | ✅ All good |
| Automated tests  | ✅ All good |
| Naming  | ✅ All good |
| Typing  | ✅ All good |
| Code comments  | ✅ All good |
| Changelog  | 👍  Almost good (left a comment) |
| Documentation  | Not applicable |
| PR description  | ✅ All good |
| Testing instructions  | ✅ All good |
| Manual testing  | Skipped |

File: plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx (hunk 14 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/payment-gateways/payment-gateways.tsx:36]: Unfortunately, we don’t have one at the moment. This will be addressed in a follow-up PR.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx (hunk 11 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/test/country-selector.test.tsx:55]: It wasn’t added intentionally, as the currently selected country is the United States, which is already displayed on the page.

File: plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts (hunk 13 above)
Comment (mordeth) [reply plugins/woocommerce/client/admin/client/settings-payments/components/country-selector/utils.ts]: Great point! I initially created the additional file because there were a few functions, but now it doesn’t seem necessary. I’ve removed it.

File: plugins/woocommerce/package.json (hunk 15 above)
Comment (mordeth) [reply plugins/woocommerce/package.json]: It seems these changes were a result of `pnpm install`. I’ve reverted them.

**Review by oaratovskyi — APPROVED** [review]
Pre-approving the PR - great work 🎉 
There's just one edge case - otherwise good to go
p1733497121285589/1733486232.454949-slack-C0800LPNCET
